test/oban/testing_test.exs

```diff
@@ -67,18 +71,46 @@ defmodule Oban.TestingTest do
     test "validating the return value of the worker's perform/1 function" do
       message = "result to be one of"
 
-      assert_perform_error(MisbehavedWorker, %{"action" => "bad_atom"}, message)
-      assert_perform_error(MisbehavedWorker, %{"action" => "bad_string"}, message)
-      assert_perform_error(MisbehavedWorker, %{"action" => "bad_error"}, message)
-      assert_perform_error(MisbehavedWorker, %{"action" => "bad_tuple"}, message)
-      assert_perform_error(MisbehavedWorker, %{"action" => "bad_snooze"}, message)
+      actions = %{
+        "bad_atom" => :bad,
+        "bad_string" => "bad",
+        "bad_error" => :error,
+        "bad_tuple" => {:ok, "bad", :bad},
+        "bad_snooze" => {:snooze, true}
+      }
+
+      for {action, result} <- actions do
+        log = """
+        Instead received:
+
+        #{inspect(result, pretty: true)}
+
+        The job will be considered a success.
+        """
+
+        assert capture_log(fn ->
+                 assert_perform_error(MisbehavedWorker, %{"action" => action}, message)
+               end) =~ log
+      end
```

Comment:
This isn't ideal. The `perform_job` function will raise an error anyhow, and I wouldn't expect to see an error. Both `perform_job` and `drain_queue` use `safe: false`. Perhaps we raise an error rather than log a warning when `safe` is `false`?

Reply:
> Perhaps we raise an error rather than log a warning when `safe` is `false`?

Yes, it's true, `perform_job/3` already raises an error if the `perform/1` returns an unexpected result, so, there is no point in adding more noise with the warning.

I just added a `maybe_log_warning/2` based on the `safe` field from the `Executor` struct :)